### PR TITLE
CP-45571: Add VM.restart_device_model function (API/CLI)

### PR DESCRIPTION
**Pull request by gangj**
Implemented as a localhost migration, can be made more efficient in future.
As it is simply a "pool_migrate", just reuse message forwarding of "pool_migrate" to handle "allowed operation" and "message routing".

### CODE REVIEW

File: ocaml/xapi/xapi_vm.ml [hunk 1]
```diff
@@ -1611,3 +1611,13 @@ let set_NVRAM_EFI_variables ~__context ~self ~value =
...
       Db.VM.set_NVRAM ~__context ~self ~value
   )
+
+let restart_device_models ~__context ~self =
+  let host = Db.VM.get_resident_on ~__context ~self in
+  (* As it is implemented as a localhost migration, just reuse message
+   * forwarding of "pool_migrate" to handle "allowed operation" and "message
+   * routing" *)
+  Helpers.call_api_functions ~__context (fun rpc session_id ->
+      Client.VM.pool_migrate ~rpc ~session_id ~vm:self ~host
```
Comment (robhoes) [inline ocaml/xapi/xapi_vm.ml:1621]: Which error will be returned if the VM is not running? It should be a bad power state.

File: ocaml/idl/datamodel_vm.ml [hunk 2]
```diff
@@ -1033,7 +1033,6 @@ let pool_migrate =
       ; Api_errors.other_operation_in_progress
       ; Api_errors.vm_is_template
       ; Api_errors.operation_not_allowed
-      ; Api_errors.vm_migrate_failed
```
Comment (robhoes) [inline ocaml/idl/datamodel_vm.ml:1036]: Why is this removed? Are you sure this error is not raised by this call?

File: ocaml/xapi/xapi_vm.ml (hunk 1 above)
Comment (gangj) [reply ocaml/xapi/xapi_vm.ml:1621]: The handler of "pool_migrate" had handled kinds of errors including the "bad power state", here is the tested result:

```
[root@lcy2-dt44 ~]# xe vm-restart-device-models vm=rhel8_202
You attempted an operation on a VM that was not in an appropriate power state at the time; for example, you attempted to start a VM that was already running. The parameters returned are the VM's handle, and the expected and actual VM state at the time of the call.
vm: 698f91f3-84f8-34e3-714e-f4cd7d8a1207 (rhel8_202)
expected: running
actual: halted
```

File: ocaml/idl/datamodel_vm.ml (hunk 2 above)
Comment (gangj) [reply ocaml/idl/datamodel_vm.ml:1036]: I copied the errors from `vm.pool_migrate` for `vm.restart_device_models`, yes, I checked the code, it is not raised from the `vm.pool_migrate` API call.

File: ocaml/xapi/xapi_vm.ml (hunk 1 above)
Comment (robhoes) [reply ocaml/xapi/xapi_vm.ml:1621]: That's good. I was worried that it would first complain about the `host` param being `OpaqueRef:NULL` (from `resident_on`).

**Review by robhoes — APPROVED** [review]

**Review by psafont — APPROVED** [review]

**Comment by robhoes** [comment]
Squashed+merged.
